storm-client/src/jvm/org/apache/storm/metrics2/TaskMetricRepo.java

```diff
@@ -34,23 +34,33 @@
     private SortedMap<String, Timer> timers = new TreeMap<>();
 
     public void addCounter(String name, Counter counter) {
-        counters.put(name, counter);
+        synchronized (this) {
```

Comment:
Looks like all the synchronization blocks are the fully body of the function, i.e. equivalent synchronized methods. If the intent is to make it finer grained, then the lock should be on "counters"  in the add....() methods and for each treeMap in the deregisterMetrics().